datalad/cmd.py

```diff
@@ -6,66 +6,152 @@
+
+        *args and **kwargs:
+           see Runner.run() and Runner.call() for available arguments.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Raises
+        ------
+        ValueError if cmd is neither a string nor a callable.
         """
+
+        if isinstance(cmd, basestring):
+            self.run(cmd, *args, **kwargs)
+        elif callable(cmd):
+            self.call(cmd, *args, **kwargs)
+        else:
+            raise ValueError("Argument 'command' is neither a string nor a callable.")
+
+    def run(self, cmd, log_stdout=False, log_stderr=True):
+        """Runs the command `cmd` using shell.
+
+        In case of dry-mode `cmd` is just added to `commands` and it is executed otherwise.
+        Allows for seperatly logging stdout and stderr  or streaming it to system's stdout
+        or stderr respectively.
+
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        log_stdout: bool
+            If True, stdout is logged. Goes to sys.stdout otherwise.
+
+        log_stderr: bool
+            If True, stderr is logged. Goes to sys.stderr otherwise.
+
+        Returns
+        -------
+        Status code as returned by the called command or `None` in case of dry-mode.
+
+        Raises
+        ------
+        RunTimeError
+           if command's exitcode wasn't 0 or None
+        """
+
+        if log_stdout:
+            outputstream = subprocess.PIPE
+        else:
+            outputstream = sys.stdout
+
+        if log_stderr:
+            errstream = subprocess.PIPE
+        else:
+            errstream = sys.stderr
+
         self.log("Running: %s" % (cmd,))
+
         if not self.dry:
-            # status, output = commands.getstatusoutput(cmd)
-            # doing manually for improved debugging
-            pipe = os.popen('{ ' + cmd + '; } 2>&1', 'r')
-            output = ''
-            for line in iter(pipe.readline, ''):
-                self.log("| " + line.rstrip('\n'), level=5)
-                output += line
-            if output[-1:] == '\n': output = output[:-1]
-            status = pipe.close() or 0
-
-            if not status in [0, None]:
-                msg = "Failed to run %r. Exit code=%d output=%s" \
-                      % (cmd, status, output)
+
+            proc = subprocess.Popen(cmd, stdout=outputstream, stderr=errstream, shell=True)
+            # shell=True allows for piping, multiple commands, etc., but that implies to not use shlex.split()
+            # and is considered to be a security hazard. So be careful with input.
+            # Alternatively we would have to parse `cmd` and create multiple subprocesses.
+
+            while proc.poll() is None:
+                if log_stdout:
+                    line = proc.stdout.readline()
+                    if line != '':
```

Comment:
yeah -- I was just whining out loud ;)  they did dump to stderr since stdout could be used for dumping actual load in the best traditions of piping unix commands... it just ruins our bloody logic for spitting at ERROR loglevel for such commands... but ok -- we will tune it up later.